### PR TITLE
NEXT-20683 - Add fastly docs

### PR DESCRIPTION
**Pull request by shyim**
Docs for https://www.fastly.com/

### CODE REVIEW

File: guides/hosting/infrastructure/reverse-http-cache.md [hunk 1]
```diff
@@ -273,5 +273,146 @@ if (obj.hits > 0) {
...
+```
+# Mitigate httpoxy application vulnerability, see: https://httpoxy.org/
+unset req.http.Proxy;
+
+# Strip query strings only needed by browser javascript. Customize to used tags.
+if (req.url ~ "(\?|&)(pk_campaign|piwik_campaign|pk_kwd|piwik_kwd|pk_keyword|pixelId|kwid|kw|adid|chl|dv|nk|pa|camid|adgid|cx|ie|cof|siteurl|utm_[a-z]+|_ga|gclid)=") {
+    # see rfc3986#section-2.3 "Unreserved Characters" for regex
+    set req.url = regsuball(req.url, "(pk_campaign|piwik_campaign|pk_kwd|piwik_kwd|pk_keyword|pixelId|kwid|kw|adid|chl|dv|nk|pa|camid|adgid|cx|ie|cof|siteurl|utm_[a-z]+|_ga|gclid)=[A-Za-z0-9\-\_\.\~]+&?", "");
+}
+set req.url = regsub(req.url, "(\?|\?&|&)$", "");
```
Comment (vagetman) [inline guides/hosting/infrastructure/reverse-http-cache.md:310]: While technically correct, this is a lot more verbose and heavy handed (use of 2 regexes) approach than necessary. You can simply use [querystring.regfilter](https://developer.fastly.com/reference/vcl/functions/query-string/querystring-regfilter/) to accomplish this.

```suggestion
if (req.url != req.url.path) {
  set req.url = querystring.regfilter(req.url, "pk_campaign|piwik_campaign|pk_kwd|piwik_kwd|pk_keyword|pixelId|kwid|kw|adid|chl|dv|nk|pa|camid|adgid|cx|ie|cof|siteurl|utm_[a-z]+|_ga|gclid");
}
```

File: guides/hosting/infrastructure/reverse-http-cache.md [hunk 2]
```diff
@@ -273,5 +273,146 @@ if (obj.hits > 0) {
...
+        unset req.http.Accept-Encoding;
+    } elsif (req.http.Accept-Encoding ~ "gzip") {
+        set req.http.Accept-Encoding = "gzip";
+    } elsif (req.http.Accept-Encoding ~ "deflate") {
+        set req.http.Accept-Encoding = "deflate";
+    } else {
+        # unkown algorithm
+        unset req.http.Accept-Encoding;
+    }
+}
```
Comment (vagetman) [inline guides/hosting/infrastructure/reverse-http-cache.md:336]: These are not necessary for Fastly. [We are normalizing `Accept-Encoding`](https://developer.fastly.com/reference/http/http-headers/Accept-Encoding/) and [we are not compressing those extensions by default](https://docs.fastly.com/en/guides/enabling-automatic-gzipping). If you still want to force `Accept-Encoding` removal for listed extensions, please use `req.url.ext`, thus something like this:

```
    if (req.url.ext ~ "^(jpg|png|gif|gz|tgz|bz2|tbz|mp3|ogg)$") {
        # No point in compressing these
        unset req.http.Accept-Encoding;
    }
```
However, I don't feel this is necessary.

File: guides/hosting/infrastructure/reverse-http-cache.md [hunk 3]
```diff
@@ -273,5 +273,146 @@ if (obj.hits > 0) {
...
+    req.method != "PATCH" &&
+    req.method != "DELETE") {
+    /* Non-RFC2616 or CONNECT which is weird. */
+    return (pass);
+}
+
+# We only deal with GET and HEAD by default
+if (req.method != "GET" && req.method != "HEAD") {
+    return (pass);
+}
```
Comment (vagetman) [inline guides/hosting/infrastructure/reverse-http-cache.md:353]: The starting point for Fastly services, [called boilerplate](https://developer.fastly.com/learning/vcl/using/#adding-vcl-to-your-service-configuration), looks like this 

```
sub vcl_recv {
#FASTLY recv

  # Normally, you should consider requests other than GET and HEAD to be uncacheable
  # (to this we add the special FASTLYPURGE method)
  if (req.method != "HEAD" && req.method != "GET" && req.method != "FASTLYPURGE") {
    return(pass);
  }
```
You can notice similar code. The snippet will be placed into the expanded `#FASTLY recv` macro, thus prior to the code. For this reason your addition might be useful, as you're doing your own `return(lookup)` and not letting it to fall through. However I'd recommend having the same code we have instead. Beside the consistency, the main reason is to allow for `FASTLYPURGE`; this is where a `PURGE` method is mapped to. That will allow a single object instant purge to work. 
```suggestion
  # Normally, you should consider requests other than GET and HEAD to be uncacheable
  # (to this we add the special FASTLYPURGE method)
  if (req.method != "HEAD" && req.method != "GET" && req.method != "FASTLYPURGE") {
    return(pass);
  }
```

File: guides/hosting/infrastructure/reverse-http-cache.md [hunk 4]
```diff
@@ -273,5 +273,146 @@ if (obj.hits > 0) {
...
+
+# Don't cache Authenticate & Authorization
+if (req.http.Authenticate || req.http.Authorization) {
+    return (pass);
+}
+
+# Always pass these paths directly to php without caching
+# Note: virtual URLs might bypass this rule (e.g. /en/checkout)
+if (req.url ~ "^/(checkout|account|admin|api)(/.*)?$") {
```
Comment (vagetman) [inline guides/hosting/infrastructure/reverse-http-cache.md:362]: You seems to be evaluating URL path here, use [req.url.path](https://developer.fastly.com/reference/vcl/variables/client-request/req-url-path/) instead
```suggestion
if (req.url.path ~ "^/(checkout|account|admin|api)(/.*)?$") {
```

File: guides/hosting/infrastructure/reverse-http-cache.md [hunk 5]
```diff
@@ -273,5 +273,146 @@ if (obj.hits > 0) {
...
+**Subroutine:** vcl_hash
+
+
+```
+# Consider Shopware http cache cookies
+if (req.http.cookie ~ "sw-cache-hash=") {
+    set req.hash += regsub(req.http.cookie, "^.*?sw-cache-hash=([^;]*);*.*$", "\1");
+} elseif (req.http.cookie ~ "sw-currency=") {
+    set req.hash += regsub(req.http.cookie, "^.*?sw-currency=([^;]*);*.*$", "\1");
+}
```
Comment (vagetman) [inline guides/hosting/infrastructure/reverse-http-cache.md:380]:
```suggestion
if (req.http.cookie:sw-cache-hash) {
  set req.hash += req.http.cookie:sw-cache-hash;
} elseif (req.http.cookie:sw-currency) {
  set req.hash += req.http.cookie:sw-currency;
}
```
In Fastly VCL we have [subfield](https://developer.fastly.com/reference/vcl/functions/miscellaneous/subfield/). No need to run regexes twice.

File: guides/hosting/infrastructure/reverse-http-cache.md [hunk 6]
```diff
@@ -273,5 +273,146 @@ if (obj.hits > 0) {
...
+```
+
+
+**Name:** Respect Shopware States to Pass the cache
+
+**Subroutine:** vcl_hit
+
+```
+if (req.http.cookie ~ "sw-states=") {
+   set req.http.states = regsub(req.http.cookie, "^.*?sw-states=([^;]*);*.*$", "\1");
```
Comment (vagetman) [inline guides/hosting/infrastructure/reverse-http-cache.md:390]: Same as the previous comment

```suggestion
if (req.http.cookie:sw-states) {
   set req.http.states = req.http.cookie:sw-states;
```

File: guides/hosting/infrastructure/reverse-http-cache.md [hunk 7]
```diff
@@ -273,5 +273,146 @@ if (obj.hits > 0) {
...
+
+   if (req.http.states ~ "logged-in" && obj.http.sw-invalidation-states ~ "logged-in" ) {
```
Comment (vagetman) [inline guides/hosting/infrastructure/reverse-http-cache.md:392]:
```suggestion
   if (std.strstr(req.http.states,"logged-in") &&  std.strstr(obj.http.sw-invalidation-states,"logged-in" ) {
```

File: guides/hosting/infrastructure/reverse-http-cache.md [hunk 8]
```diff
@@ -273,5 +273,146 @@ if (obj.hits > 0) {
...
+      return (pass);
+   }
+
+   if (req.http.states ~ "cart-filled" && obj.http.sw-invalidation-states ~ "cart-filled" ) {
```
Comment (vagetman) [inline guides/hosting/infrastructure/reverse-http-cache.md:396]:
```suggestion
   if (std.str(req.http.states,"cart-filled") && std.strstr(obj.http.sw-invalidation-states,"cart-filled" )) {
```

**Review by keulinho — APPROVED** [review]

**Comment by shyim** [comment]
@vagetman thanks for your review 💙.
I have applied your suggestions

**Pull request merged**